File: Tree/447-treeDetectionDFS.py


from collections import defaultdict
def treeDetection(left,right):
    n=len(left)
    data=defaultdict(int)
    for i in range(n):
        if left[i]!=-1:
            data[left[i]]+=1
        if right[i]!=-1:
            data[right[i]]+=1
            
    root=[x for x in range(n) if data[x]==0]
    if len(root)!=1:
        return False
    seen=set()
    def dfs(root):
        if root in seen:
            return False
        seen.add(root)
        if left[root]!=-1:
            if not dfs(left[root]):
                return False
        if right[root]!=-1:
            if not dfs(right[root]):
                return False
        return True
    return dfs(0) and n==len(seen)

left = [1, -1, 3, -1]
right = [2, -1, -1, -1]
print(treeDetection(left,right))

File: Tree/447-treeDetectionDFS.cpp
#include<vector>
#include<unordered_map>
#include<set>
#include<iostream>
#include<algorithm>
using VEC=std::vector<int>;
using MAP=std::unordered_map<int,int>;
using SEEN=std::set<int>;


bool f(const VEC& left, const VEC& right,int cur, SEEN& seen)
{
    if(std::find(seen.begin(),seen.end(),cur)!=seen.end())
        return false;
    seen.insert(cur);
    if(left[cur]!=-1)
    {
        if(!f(left,right,left[cur],seen))
            return false;
    }
    if(right[cur]!=-1)
    {
        if(!f(left,right,right[cur],seen))
            return false;
    }
    return true;
}    

bool detection(const VEC& left, const VEC& right)
{
    int n=left.size();
    MAP data{};
    for(int i=0;i<n;++i)
    {
        if(left[i]!=-1)
            data[left[i]]+=1;
        if(right[i]!=-1)
            data[right[i]]+=1;
    }
    VEC roots{};
    for(int i=0;i<n;++i)
    {
        if(data[i]==0)
            roots.push_back(i);
    }
    if(roots.size()!=1)
        return false;
    int root=roots[0];
    SEEN seen{};
    
    return f(left,right,root,seen) && seen.size()==n;
}

int main()
{
    VEC left{1, -1, 3, -1},right{2, -1, -1, -1};
    std::cout<<detection(left,right);
}